lib/EarlyFunctionAnalysis/ApplyRegisterStateDeductions.cpp

```diff
@@ -0,0 +1,81 @@
+/// \file ApplyRegisterStateDeductions.cpp
```

Comment:
I'd like to have a `librevngABI` which contains all this stuff. EFA (EarlyFunctionAnalysis) can depend on it.

I think this makes more sense from sweng POV: when a user wants to add a new ABI he know that the *only* place he has to touch is `librevngABI`.

Reply:
Actually, I feel like the `ApplyRegisterStateDeductions` doesn't actually depend on the ABI itself that much.
All it needs is all the used register lists and some basic parameters, like whether the given ABI is position-based or not.

As such, I feel like it can actually stay here. And `librevngABI` can be reserved for the time when we move ABI definitions to the runtime - it can be responsible for serializing/loading that, verification, etc. What do you think?

Reply:
Consider that by `librevngABI` I also consider `include/revng/ABI/*`.

> ApplyRegisterStateDeductions doesn't actually depend on the ABI itself that much

It is true, but we have to decide whether we want `librevngEarlyFunctionAnalysis` to know about internals of `librevngABI` (e.g., `CallingConvention::GeneralPurposeArgumentRegisters`) or viceversa (the `Yes`, `YesOrDead` thing).

Which surface is thinner?
Also consider that one day we might want to enable someone else (maybe downstream) to extend EFA results refinement with some custom criteria. In this scenario, what is the more appropriate choice?

Reply:
As for my opinion, it's easier to name things consistently (and move them to the `EarlyFunctionAnalysis` namespace), but either way, I see your point, I'll move the functions.

Reply:
Should be done (see https://github.com/revng/revng/pull/98/commits/97ebbe9a7f4ec6b4590fa3089a507dcc90f89e9f , https://github.com/revng/revng/pull/98/commits/47cd2f447966c2aa006bd65997fbf4d5d9c79467 and https://github.com/revng/revng/pull/98/commits/32c2d3be5c29afb9b444c2cbf646fac2bc5ba061).